src/intersections/lines.jl

```diff
@@ -43,14 +28,30 @@ end
 function intersectpoint(l1::Line, l2::Line)
   a, b = l1(0), l1(1)
   c, d = l2(0), l2(1)
-
-  v1  = a - b
-  v2  = c - d
-  v12 = a - c
-
-  # the intersection point lies in between a and b at a fraction s
-  # (https://en.wikipedia.org/wiki/Line-line_intersection#Formulas)
-  s = (v12[1] * v2[2] - v12[2] * v2[1]) / (v1[1] * v2[2] - v1[2] * v2[1])
-
-  a - s*v1
+  p₁, _ = intersectparameters(a, b, c, d)
+  a + p₁ * (b - a)
+ end
+
+# compute the parameters of the lines defined by the points a -> b and c -> d 
+# so that the resulting points have minimal distance
+# the ranks help to identify the different types of intersection
+function intersectparameters(a::Point{N,T}, b::Point{N,T}, c::Point{N,T}, d::Point{N,T}) where {N,T}
+  # solves the equation (approximately): a + λ₁ ⋅ v⃗₁ = c + λ₂ ⋅ v⃗₂, v⃗₁ = b - a, v⃗₂ = d - c
+  A = [(b - a) (c - d)]
+  b = c - a
+
+  λ = A \ b
+
+  # calculate the rank of the augmented matrix
+  rₐ = rank([A b], atol = atol(T))
+  # calculate the rank of the rectangular matrix
+  r = rank(A, atol = atol(T))
```

Comment:
I will think about this, a benchmark might be helpful to decide.

Reply:
So I tested the approach to calculate the QR decomposition (LU is instable if matrix is singular).
Gaussian elimination (of an upper triangular matrix) is only needed if lines are not parallel or colinear (`r == 2`), (otherwise paramters are set to `zero(T)`).

For the evaluation i need only submatrices of the `Q` and `R` matrices. When accessing them, they are converted back from SArray to normal arrays which makes to evaluation slow (`intersectparameters_3`)
Therefore I implemented an additional QR decomposition on the static matrix A alone, which in total seems to be faster and commited that version  (`intersectparameters_2`) - maybe you have a better solution

![image](https://user-images.githubusercontent.com/67096719/179749737-a2c689af-46c7-4699-8f97-dacfa37ec7eb.png)

intersectparameters_2:
``` code
function intersectparameters_2(a::Point{N,T}, b::Point{N,T}, 
                             c::Point{N,T}, d::Point{N,T}) where {N,T}
  # solves the equation (approximately): 
  # a + λ₁ ⋅ v⃗₁ = c + λ₂ ⋅ v⃗₂, 
  # with v⃗₁ = b - a, v⃗₂ = d - c
  A = [(b - a) (c - d)]
  y = c - a
  QR = qr([A y])

  # calculate the rank of the augmented matrix
  rₐ = sum(abs.(diag(QR.R)) .> atol(T))
  # calculate the rank of the rectangular matrix
  r = sum(abs.(diag(QR.R)[1:2]) .> atol(T))

  if r >= 2 # calculate parameters of intersection or closest point
    QR = qr(A)
    λ = QR.R \ QR.Q' * (c-a)
    return λ[1], λ[2], r, rₐ
  else # parallel or collinear
    return zero(T), zero(T), r, rₐ
  end
  
  # Intersection r == rₐ == 2
  # Collinear: r == rₐ == 1
  # no intersection: r != rₐ
  #   no intersection and parallel:  r == 1, rₐ == 2
  #   no intersection, skew lines: r == 2, rₐ == 3
end
```

intersectparameters_3:
```
function intersectparameters_3(a::Point{N,T}, b::Point{N,T}, 
                             c::Point{N,T}, d::Point{N,T}) where {N,T}
  # solves the equation (approximately): 
  # a + λ₁ ⋅ v⃗₁ = c + λ₂ ⋅ v⃗₂, 
  # with v⃗₁ = b - a, v⃗₂ = d - c
  A = [(b - a) (c - d)]
  y = c - a
  QR = qr([A y])

  # calculate the rank of the augmented matrix
  rₐ = sum(abs.(diag(QR.R)) .> atol(T))
  # calculate the rank of the rectangular matrix
  r = sum(abs.(diag(QR.R)[1:2]) .> atol(T))

  if r >= 2 # calculate parameters of intersection or closest point
    λ = QR.R[1:2,1:2] \ (QR.Q' * y)[1:2] 
    return λ[1], λ[2], r, rₐ
  else # parallel or collinear
    return zero(T), zero(T), r, rₐ
  end
  
  # Intersection r == rₐ == 2
  # Collinear: r == rₐ == 1
  # no intersection: r != rₐ
  #   no intersection and parallel:  r == 1, rₐ == 2
  #   no intersection, skew lines: r == 2, rₐ == 3
	
end
```